### PR TITLE
Added factor method to FiniteSet (Non expr class)

### LINKED ISSUE #18064: Add .evalf() method to sympy.physics.vector.Vector and Dyadic
These two classes are missing many of the core SymPy methods because they do not inherit from the core SymPy classes. It would be nice to add them, starting with `.evalf()`.

### PR DESCRIPTION
**Pull request by mohitacecode**
I have added factor method as mentioned in the issue #18064

#### References to other Issues or PRs
fixes #18064 
#### Brief description of what is fixed or changed
Added factor method to FiniteSet (Non expr class)

#### Release Notes

*   sets
     *  Added `factor` method to `FiniteSet` (Non expr class)

### CODE REVIEW

File: sympy/sets/sets.py [hunk 1]
```diff
@@ -1765,6 +1765,16 @@ class FiniteSet(Set, EvalfMixin):
...
     is_finite_set = True
 
+    def factor(self, *args, **kwargs):
+        """
+        Implements the SymPy factor routine, on the scalar parts
+        of a basis-dependent expression.
+        factor's documentation
+        ========================
+        """
+        fctr_components = [fctr(v, *args, **kwargs) * k for k, v in self.components.items()]
```
Comment (divyanshu132) [inline sympy/sets/sets.py]: I can't find any existing method named `fctr`.

File: sympy/sets/sets.py [hunk 2]
```diff
@@ -1765,6 +1765,16 @@ class FiniteSet(Set, EvalfMixin):
...
+        return self._add_func(*fctr_components)
+
```
Comment (divyanshu132) [inline sympy/sets/sets.py]: You have trailing whitespace look at [this](https://travis-ci.org/sympy/sympy/jobs/635934071?utm_medium=notification&utm_source=github_status)

**Review by divyanshu132 — COMMENTED** [review]
Also, it can have a release note entry

File: sympy/sets/sets.py (hunk 1 above)
Comment (mohitacecode) [reply sympy/sets/sets.py]: I just fixed this error and also trailing space one now some new error is coming :(

=>ImportError: cannot import name 'sqrt' from partially initialized module 'sympy'

Comment (mohitacecode) [reply sympy/sets/sets.py]: do i have make another PR for it?

File: sympy/sets/sets.py [hunk 3]
```diff
@@ -26,9 +26,8 @@
 from sympy.utilities.exceptions import SymPyDeprecationWarning
 from sympy.utilities.iterables import iproduct, sift, roundrobin
 from sympy.utilities.misc import func_name, filldedent
-
 from mpmath import mpi, mpf
-
+from sympy.vector.basisdependent import factor as fctr
```
Comment (oscarbenjamin) [inline sympy/sets/sets.py]:
```
from sympy.polys import factor
```
No need to rename as `fctr`

Comment (mohitacecode) [reply sympy/sets/sets.py]: I tried using `from sympy.polys import factor` but it shows this error "cannot import name 'factor' from partially initialized module 'sympy.polys".

The previous one was working but it start giving a new issue in different file "ImportError: cannot import name 'sqrt' from partially initialized module 'sympy' ".

Comment (oscarbenjamin) [reply sympy/sets/sets.py]: Maybe you'll need to use a delayed import then and import it in the function. You shouldn't be importing it from vector as it isn't defined there.

Try this:
https://github.com/sympy/sympy/blob/f119ab5d30e8b373dc8956d41b7c11e53c140209/sympy/core/expr.py#L3583-L3586

File: sympy/sets/sets.py (hunk 1 above)
Comment (divyanshu132) [reply sympy/sets/sets.py]: No, need to make a separate PR try importing `factor` inside the function as oscar has suggested.

**Comment by oscarbenjamin** [comment]
This will need tests

**Comment by oscarbenjamin** [comment]
If you run `bin/test quality` before committing it will pick up on these issues with whitespace a lot faster than pushing here. You can also run `flake8 sympy` (after installing `flake8`).

File: sympy/sets/sets.py (hunk 3 above)
Comment (mohitacecode) [reply sympy/sets/sets.py]: thank you so much this works :)

File: sympy/sets/sets.py (hunk 1 above)
Comment (mohitacecode) [reply sympy/sets/sets.py]: thanks so much :)

**Comment by namannimmo10** [comment]
Try it this way:
```
<!-- BEGIN RELEASE NOTES -->
*   sets
     *  Added `factor` method to `FiniteSet` (Non expr class)
<!-- END RELEASE NOTES -->
```

**Comment by asmeurer** [comment]
What's the purpose of having this? I don't think the sets should inherit every method from Expr.

**Comment by oscarbenjamin** [comment]
We need to decide how this sort of thing should be handled so that it works reliably. The `factor` function works on `FiniteSet` but there is no `factor` method:
```julia
In [13]: s = FiniteSet(x**2 + x)                                                                                                               

In [14]: s                                                                                                                                     
Out[14]: 
⎧ 2    ⎫
⎨x  + x⎬
⎩      ⎭

In [15]: factor(s)                                                                                                                             
Out[15]: {x⋅(x + 1)}

In [16]: s.factor()                                                                                                                            
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-16-0e76aed4bd64> in <module>
----> 1 s.factor()

AttributeError: 'FiniteSet' object has no attribute 'factor'
```
In general I think that `Expr` (and `Basic`) have too many methods and I don't want to add more. I do want things to be more consistent and predictable though.

**Comment by mohitacecode** [comment]
@oscarbenjamin Can you please review the test I have added

**Comment by oscarbenjamin** [comment]
It looks like you've added a test in #18375 for this PR. The test should go in this PR with the code that is changed because it will fail otherwise.

**Comment by mohitacecode** [comment]
I've just added test to test_sets.py and for the change in this file I have to open a new pr can you tell me how to add test to this pr?

**Comment by mohitacecode** [comment]
@oscarbenjamin  I have added a test for this pr.

**Comment by oscarbenjamin** [comment]
I'm not sure whether we want this method...

**Comment by mohitacecode** [comment]
Where are you struggling?

**Comment by czgdp1807** [comment]
Any updates on this PR?

**Comment by czgdp1807** [comment]
@mohitacecode Any news?

**Pull request closed without merging**